### ISSUE TITLE
Only using the Gradient and Self define iteration with mxnet python API

### ISSUE THREAD

**Issue by toothacher17** [dmlc/mxnet#369]
Hi all,

I am building a model which integrates some NN layers into a generative model and uses EM to do MLE. **I do not have "label" and "loss" function, the error is implicit.**

So I am not going to use mxnet "traditionally". **My basic idea is I would need the gradients of some layer in the NN, which I think could be provided by mxnet.** With these gradients and some other self defined gradients, I could do backprop and batch SGD then. Sample in E step, and maximize in M step.

I looked given examples and the symbolic configuration documentations. I would say it is helpful and detailed. However, I am still wondering that if mxnet could help implement my idea? Or I should try the cpp interfaces?

My basic steps:

1. data, store in NDArray:
data = mx.NDArray()

2. some symbolic layers given by mxnet, only need gradients
A = mx.sym.Layer(params)

3. self defined layer: self defined, only need gradients
def some_f(data, w)

4. iteration (self defined or using **Data Iterator**?)
E: sample new state, M: backprop and SGD

So the questions could be that is it possible to implement step2 and step4 with mxnet, using the symbolic configuration or Data Iterator?  Or more general, is it feasible to self define the iterator and symbol layer?

If you think this question is not so interesting or you are feeling "you are just implementing my ideas", please tell me so I would try some other methods. **Basically, I would only want the mxnet to provide with some descents and try to create iterators with mxnet.** I think this need would be general for those who tried to combine NN with some other models. 

Really thanks a lot for your attention. Any suggestions for implementation could be really helpful.

**Comment by tqchen** [comment]
not sure I fully understand your question. Here are some answers

- It is definitely possible to do customized data input, you can either use io.NDArrayIter, or simply feed your data into executor, if you are using your own training functions.
- It is possible to backprop the gradient back and get the gradient of arrays, you can do ```executor.backward(out_grads=output gradient)```

So what you can likely do is:
- Bind an executor using the set of symbols, to define your computation.
- Do a  executor.forward to get output
- Do some of your EM tricks or other tricks to get the output gradient
- Call executor.backward to prop gradient back, and get the gradient you need
- Do the update you want in python.

**Comment by tqchen** [comment]
In terms of self defined layer. It depends on whether what you need can be readily composed by mxnet's current symbolic API, or if there is more you need. You are always welcomed to open another issue and tell us about your usecase. Thanks!

**Comment by toothacher17** [comment]
Great to see your reply! Fast and informative!  Great, with your reply, I think I could start to make progress now, specifically:

1. yes , I would need to customize my data input, so I would feed into an executor

2. great! I would use the "executor.backward" to get the gradients and see the documents in more details

3. For the self defined layer, no, I do not think it could be composed by mxnet's current symbolic API. **Basically, it is not a "layer", just I would use the executor forward result in some other complex function in this "layer", and then use its gradients to combine with executor.backward to complete the chain rule**. Can a python function combine with forward and backward result? **So I see the doc says it could return with NDArray, so I could directly use the forward and backward (gradients) result, right?**

4. If I can get the forward and backward result in one iteration, then E and M step could be done easily, I could **write my own iteration in python rather than fit into mxnet framework** as you mentioned, right? 

5. Thanks a lot. After my implementation, I would be very happy to provide details about the usecase.

I would continue to arise discussions and asked some questions here. I would now check about the document as you suggested. Thanks a lot

**Comment by tqchen** [comment]
Yes, you can directly play with the executor interface and do things you like in NDArray, and training customization and optimization you like. 

There are some modules such as initializer and optimizer which you might be able to re-use, but not required.

**Comment by antinucleon** [comment]
You may find this example useful.
https://github.com/dmlc/mxnet/blob/master/example/adversary/adversary_generation.ipynb

In this example, there is customized data, layer, loss, backprop, forward etc

**Comment by tqchen** [comment]
I am going to close this for now, please feel free to re-open if you have future questions on this or want to share more of your experience with the community

**Comment by toothacher17** [comment]
Hi, 

Thanks a lot for your help and the example you showed is detailed and helpful in some ways. And I am "almost" finish the python implementation. Basically:

1. configure neural nets
2. bind those symbols to an executor
3. self defined iteration and EM update
(3.1) E  step: executor.forward and combine some functions, then sampling
(3.2) M step: executor.backward and combine some functions, then updating

Are these steps right? However, current python implementation is not efficient enough, considering the python sampler. So my advisor suggested to the cpp implementation, and the sampler could use the "dmlc/experimental-lda" (need many changes). So I was wondering if you have cpp examples? I saw the example folder and did not see any cpp examples

I was delayed by something else in the past weeks, and would try to catch up the progress in the coming weeks, so please do not close this talk recently.

Thanks  a lot for your help

**Comment by toothacher17** [comment]
Hi,

Are there any interesting implementation examples in cpp? Or I should read the cpp API interface instead?

Thanks a lot

**Comment by tqchen** [comment]
for speed reasons, normally you only need to implement the Operator using C++ and compose them with python

**Comment by toothacher17** [comment]
Thanks a lot for the suggestion. 

Basically, I did not need to build a whole new layer. Since the nice design of symbolic and binding executors, I just need to implement the outer gradient for the executor to do backward. The whole model also used the dmlc/experiment-lda, which is a fast sampler. The whole project would be better in cpp for speed. 

By the way, another question is sparse input feature for neural nets. I use the neural nets for modeling text and using side information as feature input. Therefore, the input is very sparse. I use NDArray as input, so it would be a very sparse matrix from the view of numpy array. Is this okay and efficient to use NDArrary? Or it has some better way to do so?

**Comment by tqchen** [comment]
There is a special operator called embedding that takes sequence of text. We will indeed need a sparse matrix dot product operator

**Comment by toothacher17** [comment]
Yes sometimes the feature input size could be very large but also very sparse for modeling text. Not only the spare matrix dot product operator, but also the NDArray(maybe a special SparseNDArray)? 

I just configured the symbols and bind it with the NDArray as feature input. Storing in such sparse data, (like index map, if the index existed, then get 1 for the index) with numpy matrix could be also be waste of memory.

**Comment by tqchen** [comment]
Let us move the sparse matrix support discussion to https://github.com/dmlc/mxnet/issues/773
